contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,193 @@
+    // +optional
+    Preferred []WeightedTopologyTerm
+}
+
+// Defines a topolgoy level.
+type TopologyTerm struct {
+    // A topology key, associated with the corresponding weight.
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // The built-in topology keys can be good examples and we recommend users switch to a similar mode for portability, but it's NOT enforced.     
+    // Users can define whatever topolgoy key they like since topology is arbitrary.
+    // Required.
+    TopologyKey string `json:"topologyKey" protobuf:"bytes,1,opt,name=topologyKey"`
+}
+
+// The weights of all of the matched WeightedTopologyTerm fields are added to find the most preferred service endpoints.
+type WeightedTopologyTerm struct {
+    // weight associated with matching the corresponding topologyKey,
+    // in the range 1-100. Default Weight value is 1 instead of 0!
+    Weight int32 `json:"weight" protobuf:"varint,1,opt,name=weight"`
+    // A topology key, associated with the corresponding weight.
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // The built-in topology keys can be good examples and we recommend users switch to a similar mode for portability, but it's NOT enforced.     
+    // Users can define whatever topolgoy key they like since topology is arbitrary.
+    // Required.
+    TopologyKey string `json:"topologyKey" protobuf:"bytes,2,opt,name=topologyKey"`
+}
+```
+
+An example of service that uses topologyKey:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: with-service-topology
+spec:
+  required:
+    - topologyKey: failure-domain.beta.kubernetes.io/zone
+    - topologyKey: failure-domain.beta.kubernetes.io/region
+  preferred:
+    - weight: 100
+      topologyKey: kubernetes.io/hostname
+```
+
+In our example, both "same zone" AND "same region" are required while "same host" is not required but preferred.
+
+#### Endpoints API changes
+
+Although `NodeName` was already added to `EndpointAddress`, we want `Endpoints` to carry more node's topology informations so that allowing more topology levels other than hostname. 
+
+So, create a new `NodeLabels` field in `Endpoints.Subsets.Addresses` for identifying what topology domain the endpoints pod exists, e.g. what host, rack, zone, region etc. In other words, copy the labels of node hosting the endpoint to `EndpointAddress.NodeLabels`.
+
+```go
+type EndpointAddress struct {
+  // labels of node hosting the endpoint
+  NodeLabels map[string]string
```

Comment:
since endpoints addresses can point to things not hosted on nodes that would still benefit from topology, should this be more generic, e.g. `Topology map[string]string`

Reply:
make sense.